# Ejercicios Pair Programming 
## Lección 10 - EDA I

Empezamos una nueva lección, y este es el momento de poner en práctica los conocimientos adquiridos del Análisis Exploratorio de Datos. Si recordamos, en la lecciones de Pandas estuvimos usando un dataset sobre ataques de tiburones, y que hicimos algunos cambios a lo largo de los ejercicios, incluimos filas, columnas e incluso limpiamos algunas colummas 💪🏽! Para todas los ejercicios de pair programming de EDA I y II usaremos el dataframe generado en el ejercicio de Pandas VII.

En el ejercicio de hoy nos familiarizaremos con el dataset completo sobre el que trabajaremos en la parte de limpieza.

Pero para hacer una exploración buena del dataset no lo podemos hacer como pollo sin cabeza a ver que nos encontramos. Necesitamos hacernos una serie de preguntas o hipótesis que queramos contestar. En este caso os plantemos las siguientes preguntas a contestar a lo largo de los ejercicios de pair programming de este módulo:

### Hipótesis

Las hipótesis que os planteamos son:
- ¿Es Australia el sitio más peligroso y letal para estar relajada en la playa?
- ¿Hay diferencias entre los países en los ataques a hombres y mujeres?
- ¿Cuáles son las edades que más sufren ataques?
- Independientemente de la edad, ¿sufren los hombres más ataques que las mujeres?
- ¿En qué mes ocurren más ataques?
- ¿Cuál es la relación entre la especie y el tipo de ataque (si es fatal o no)?
- ¿Cómo han evolucionado los ataques a lo largo del tiempo?

In [23]:
import pandas as pd

In [24]:
pd.options.display.max_columns = None

In [25]:
df = pd.read_csv("attacks4.csv", index_col = 0)

In [26]:
df.head(2)

case_number    year        type country        area  \
0    2018.06.25  2018.0     Boating     USA  California   
1  2018.06.03.a  2018.0  Unprovoked  BRAZIL  Pernambuco   

                      location  activity                    name age  \
0  Oceanside, San Diego County  Paddling             Julie Wolfe  57   
1        Piedade Beach, Recife  Swimming  Jose Ernesto da Silva   18   

             time     species_  \
0           18h00  White shark   
1  Late afternoon  Tiger shark   

                                                href      siglo  unnamed:_0  \
0  http://sharkattackfile.net/spreadsheets/pdf_di...  siglo-XXI           0   
1  http://sharkattackfile.net/spreadsheets/pdf_di...  siglo-XXI           6   

                                              injury  mes fatal sex  
0  No injury to occupant, outrigger canoe and pad...  Jun     N   F  
1                                              FATAL  Jun     Y   M

### Objetivos

Los objetivos de hoy son:
1. Haced una exploración inicial del dataframe:

- ¿Cuál es el número de filas y columnas?

In [27]:
print(f"El número de filas de nuestro DataFrame es {df.shape[0]} y el número de columnas es {df.shape[1]}")

El número de filas de nuestro DataFrame es 1672 y el número de columnas es 18


- ¿Cuál es el porcentaje de valores nulos en cada columna? Presenta los resultados en un dataframe, donde las columnas sean el nombre de las variables y el porcentaje de nulos.

In [28]:
df_porcentaje = pd.DataFrame(df.isnull().sum()*100/df.shape[0]).T
df_porcentaje

case_number  year  type   country      area  location  activity      name  \
0          0.0   0.0   0.0  0.598086  2.751196  3.050239  1.854067  0.538278   

        age      time  species_  href  siglo  unnamed:_0    injury        mes  \
0  9.210526  9.569378  7.535885   0.0    0.0         0.0  0.478469  13.038278   

      fatal       sex  
0  5.801435  0.897129

- ¿Qué tipos de datos hay en cada columna?

In [29]:
df.dtypes

case_number     object
year           float64
type            object
country         object
area            object
location        object
activity        object
name            object
age             object
time            object
species_        object
href            object
siglo           object
unnamed:_0       int64
injury          object
mes             object
fatal           object
sex             object
dtype: object

- Extraed la información general del dataframe.

In [30]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 1672 entries, 0 to 1671
Data columns (total 18 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   case_number  1672 non-null   object 
 1   year         1672 non-null   float64
 2   type         1672 non-null   object 
 3   country      1662 non-null   object 
 4   area         1626 non-null   object 
 5   location     1621 non-null   object 
 6   activity     1641 non-null   object 
 7   name         1663 non-null   object 
 8   age          1518 non-null   object 
 9   time         1512 non-null   object 
 10  species_     1546 non-null   object 
 11  href         1672 non-null   object 
 12  siglo        1672 non-null   object 
 13  unnamed:_0   1672 non-null   int64  
 14  injury       1664 non-null   object 
 15  mes          1454 non-null   object 
 16  fatal        1575 non-null   object 
 17  sex          1657 non-null   object 
dtypes: float64(1), int64(1), object(16)
memory usage

2. Cread dos dataframes nuevos, uno que incluya solo las variables categóricas y otro que incluya solo las variables numéricas.

In [31]:
df_categoricas = pd.DataFrame(df.select_dtypes(include = "object"))
df_categoricas.head(3)

case_number        type country        area                      location  \
0    2018.06.25     Boating     USA  California   Oceanside, San Diego County   
1  2018.06.03.a  Unprovoked  BRAZIL  Pernambuco         Piedade Beach, Recife   
2  2018.05.26.b  Unprovoked     USA     Florida  Cocoa Beach, Brevard  County   

   activity                    name age            time        species_  \
0  Paddling             Julie Wolfe  57           18h00     White shark   
1  Swimming  Jose Ernesto da Silva   18  Late afternoon     Tiger shark   
2   Walking               Cody High  15           17h00  Bull shark, 6'   

                                                href      siglo  \
0  http://sharkattackfile.net/spreadsheets/pdf_di...  siglo-XXI   
1  http://sharkattackfile.net/spreadsheets/pdf_di...  siglo-XXI   
2  http://sharkattackfile.net/spreadsheets/pdf_di...  siglo-XXI   

                                              injury  mes fatal sex  
0  No injury to occupant, outrigger canoe and pad...  Jun     N   F  
1                                              FATAL  Jun     Y   M  
2                              Lower left leg bitten  May     N   M

In [32]:
df_categoricas.describe()

case_number        type country     area  \
count         1672        1672    1662     1626   
unique        1672           6      95      236   
top     2018.06.25  Unprovoked     USA  Florida   
freq             1        1432     741      346   

                                location activity  name   age   time  \
count                               1621     1641  1663  1518   1512   
unique                              1310      404  1536    94    248   
top     New Smyrna Beach, Volusia County  Surfing  male    19  11h00   
freq                                  73      426    97    70     74   

           species_                                               href  \
count          1546                                               1672   
unique          856                                               1672   
top     White shark  http://sharkattackfile.net/spreadsheets/pdf_di...   
freq             76                                                  1   

           siglo injury   mes fatal   sex  
count       1672   1664  1454  1575  1657  
unique         2   1173    12     2     2  
top     siglo-XX  FATAL   Aug     N     M  
freq         887    154   153  1258  1435

Otra forma de obtener el dataframe de los estadísticos de las columnas categóricas es hacerlo directamente así:

In [33]:
df_categoricas2 = pd.DataFrame(df.describe(include = "object"))
df_categoricas2

case_number        type country     area  \
count         1672        1672    1662     1626   
unique        1672           6      95      236   
top     2018.06.25  Unprovoked     USA  Florida   
freq             1        1432     741      346   

                                location activity  name   age   time  \
count                               1621     1641  1663  1518   1512   
unique                              1310      404  1536    94    248   
top     New Smyrna Beach, Volusia County  Surfing  male    19  11h00   
freq                                  73      426    97    70     74   

           species_                                               href  \
count          1546                                               1672   
unique          856                                               1672   
top     White shark  http://sharkattackfile.net/spreadsheets/pdf_di...   
freq             76                                                  1   

           siglo injury   mes fatal   sex  
count       1672   1664  1454  1575  1657  
unique         2   1173    12     2     2  
top     siglo-XX  FATAL   Aug     N     M  
freq         887    154   153  1258  1435

In [34]:
df_numericas = pd.DataFrame(df.describe()).T
df_numericas

count         mean          std  min      25%     50%     75%  \
year        1672.0  1832.688397   528.969901  0.0  1964.00  1997.0  2008.0   
unnamed:_0  1672.0  2574.174043  1756.622536  0.0  1218.25  2210.0  3687.5   

               max  
year        2018.0  
unnamed:_0  6285.0

- ¿Cuáles son los principales estádisticos de los dataframes?

De las variables numéricas sólo tenemos la columna 'year', y sus principales estadísticos tienen un valor de:
- Hay un total de 1672 filas.
- La media es de 1832,68 años.
- La desviación estándar es de 528,96 con respecto a la media.
- El año menor es el 0 y el año mayor es el 2018.
- El cuartil 1 es el año 1964 (el 25 % de los datos es inferior a 1964)
- El cuartil 2 o mediana es el año 1997.
- El cuartil 3 es el año 2008.

De las variables categóricas tenemos el resto de las columnas y sus principales estadísticos son:
- Número total de filas.
- Los valores únicos de cada columna.
- Top, que es la moda, valor que más se repite.
- Y la frecuencia de dicho top.

- Extraed los valores únicos su frecuencia. BONUS presenta los resultados en un dataframe.

Donde variable es cada una de las variables categóricas que tenemos en el dataset y n_unicos es una lista con los nombres de los valores únicos de cada variable.

Valores únicos y su frecuencia

In [35]:
#frecuencia
for col in df_categoricas.columns:
    print("--------------")
    print(f"En la columna '{col}' tenemos los siguientes valores:")
    print(pd.DataFrame(df[col].value_counts()))
    print("--------------")

--------------
En la columna 'case_number' tenemos los siguientes valores:
              case_number
2018.06.25              1
1975.07.15              1
1979.11.27              1
1980.00.00.b            1
1980.01.10              1
...                   ...
2005.10.13              1
2005.10.19              1
2005.10.21              1
2005.10.22              1
ND.0001                 1

[1672 rows x 1 columns]
--------------
--------------
En la columna 'type' tenemos los siguientes valores:
              type
Unprovoked    1432
Provoked       115
Invalid         92
Boating         17
Sea Disaster    12
Boat             4
--------------
--------------
En la columna 'country' tenemos los siguientes valores:
                    country
USA                     741
AUSTRALIA               343
SOUTH AFRICA            234
BAHAMAS                  29
NEW ZEALAND              21
...                     ...
GREENLAND                 1
ST. MAARTIN               1
UNITED KINGDOM            1
SWEDEN

El output de las frecuencias es un dataframe, pero printeado en forma de serie en el cual vemos por cada valor único el número de veces que se repite.

Bonus

In [36]:
valores_unicos = {col:df[col].unique().tolist() for col in df_categoricas.columns}

In [37]:
df_2 = pd.DataFrame(columns=["variable", "n_uniques"])
df_2["variable"] = valores_unicos.keys()
df_2["n_uniques"] = valores_unicos.values()

In [38]:
df_2

variable                                          n_uniques
0   case_number  [2018.06.25, 2018.06.03.a, 2018.05.26.b, 2018....
1          type  [Boating, Unprovoked, Provoked, Invalid, Sea D...
2       country  [USA, BRAZIL, AUSTRALIA, ENGLAND, COSTA RICA, ...
3          area  [California, Pernambuco, Florida, Queensland, ...
4      location  [Oceanside, San Diego County, Piedade Beach, R...
5      activity  [Paddling, Swimming, Walking, Feeding sharks, ...
6          name  [Julie Wolfe, Jose Ernesto da Silva , Cody Hig...
7           age  [57, 18, 15, 32, 21, 30, 60, 33, 19, 25, 10, 6...
8          time  [18h00, Late afternoon, 17h00, Morning, 08h15,...
9      species_  [White shark, Tiger shark, Bull shark, 6', Gre...
10         href  [http://sharkattackfile.net/spreadsheets/pdf_d...
11        siglo                              [siglo-XXI, siglo-XX]
12       injury  [No injury to occupant, outrigger canoe and pa...
13          mes  [Jun, May, Apr, Mar, nan, Feb, Jan, Dec, Nov, ...
14        fatal                                        [N, Y, nan]
15          sex                                        [F, M, nan]

In [39]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 1672 entries, 0 to 1671
Data columns (total 18 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   case_number  1672 non-null   object 
 1   year         1672 non-null   float64
 2   type         1672 non-null   object 
 3   country      1662 non-null   object 
 4   area         1626 non-null   object 
 5   location     1621 non-null   object 
 6   activity     1641 non-null   object 
 7   name         1663 non-null   object 
 8   age          1518 non-null   object 
 9   time         1512 non-null   object 
 10  species_     1546 non-null   object 
 11  href         1672 non-null   object 
 12  siglo        1672 non-null   object 
 13  unnamed:_0   1672 non-null   int64  
 14  injury       1664 non-null   object 
 15  mes          1454 non-null   object 
 16  fatal        1575 non-null   object 
 17  sex          1657 non-null   object 
dtypes: float64(1), int64(1), object(16)
memory usage

### Conclusiones

De qué nos sirven estos números y dataframe? Extraed conclusiones de ellos.

Por ejemplo:
- ¿Hay alguna columna que no tenga sentido su tipo de datos?

Visualizando el tipo de datos de todas nuestras columnas comprobamos que la columna "age" es de tipo object y debería ser "int64" ya que la edad es entera. También en el caso de la columna "year" que es de tipo float se podría reemplazar su tipo de datos a "int64" ya que los años son enteros.

- ¿Hay alguna columna que no tenga sentido su contenido?

Unnamed:_0 no nos sirve para ningún análisis en nuestro dataframe por lo que sería conveniente eliminarla.

href: no nos sirve en nuestro análisis porque son enlaces para páginas webs.

location: no nos serviría porque es un espacio muy detallado y para el análisis no nos valdría.

time: su contenido en algunas filas es la hora exacta y en otras es como un rango de tiempo, por lo que necesitaríamos unificar el contenido para que se puede analizar.

- ¿Hay alguna columna que tenga demasiadas categorías o valores que nos pueda hacer difícil trabajar con ella?

species_ = esta columna tiene muchas categorías y es imposible poder trabajar con ella a no ser que las agrupemos en categorías más generales.

name = es bastante complicado trabajar con esta columna porque son valores únicos y podríamos prescindir de ella.

injury = especifica demasiado el ataque y al tener la columna fatal ya no nos sirve dicha columna para el análisis porque así solo nos centraríamos en los ataques mortales y los que no.

Happy coding 🦈